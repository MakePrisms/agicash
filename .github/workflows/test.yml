name: Preview Deployment

on:
  pull_request:
    types: [opened, synchronize]

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - uses: supabase/setup-cli@v1
        with:
            version: 2.6.8

      - name: Testing
        env:
          SUPABASE_ACCESS_TOKEN: ${{ secrets.SUPABASE_ACCESS_TOKEN }}
          SUPABASE_PROJECT_ID: ${{ vars.SUPABASE_PROJECT_ID }}
        run: supabase branches list --experimental --project-ref $SUPABASE_PROJECT_ID

      # Get connection details for the automatically created branch
      - name: Get Supabase Branch Connection Details
        env:
          SUPABASE_ACCESS_TOKEN: ${{ secrets.SUPABASE_ACCESS_TOKEN }}
          SUPABASE_PROJECT_ID: ${{ vars.SUPABASE_PROJECT_ID }}
        run: |
          # Get branch name from PR
          BRANCH_NAME="${GITHUB_HEAD_REF}"
          BRANCH_ID=$(supabase branches list --experimental --project-ref $SUPABASE_PROJECT_ID | sed -n "s/.*\([0-9a-f]\{8\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{12\}\).*$BRANCH_NAME.*/\1/p")
          
          # Extract and set connection details as environment variables
          echo "BRANCH NAME: $BRANCH_NAME"
          echo "BRANCH_ID: $BRANCH_ID"
          supabase branches get $BRANCH_ID --project-ref $SUPABASE_PROJECT_ID --experimental --output json
        #   echo "SUPABASE_URL=$(echo $BRANCH_CONFIG | jq -r '.url')" >> $GITHUB_ENV
        #   echo "SUPABASE_ANON_KEY=$(echo $BRANCH_CONFIG | jq -r '.anon_key')" >> $GITHUB_ENV