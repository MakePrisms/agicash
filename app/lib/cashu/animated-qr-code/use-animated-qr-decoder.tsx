import { URDecoder } from '@ngraveio/bc-ur';
import { useEffect, useRef, useState } from 'react';

type Props = {
  fragment: string;
  onDecode?: (decodedText: string) => void;
};

export function useAnimatedQRDecoder({ fragment, onDecode }: Props) {
  const [progress, setProgress] = useState(0);
  const [error, setError] = useState<Error | null>(null);

  const decoder = useRef<URDecoder | null>(null);
  decoder.current = new URDecoder();

  useEffect(() => {
    if (!decoder.current || fragment === '') return;

    try {
      decoder.current.receivePart(fragment);
      setProgress(decoder.current.estimatedPercentComplete() || 0);

      if (decoder.current.isComplete() && decoder.current.isSuccess()) {
        const ur = decoder.current.resultUR();
        const decoded = ur.decodeCBOR().toString('utf-8');
        onDecode?.(decoded);
      }
    } catch (e) {
      console.error('Failed to decode QR fragment', e);
      setError(
        e instanceof Error ? e : new Error('Failed to decode QR fragment'),
      );
    }
  }, [fragment, onDecode]);

  return {
    progress,
    error,
  };
}
