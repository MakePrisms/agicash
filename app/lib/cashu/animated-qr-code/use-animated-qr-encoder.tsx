import { UR, UREncoder } from '@ngraveio/bc-ur';
import { useRef } from 'react';
import { useInterval } from 'usehooks-ts';

type Props = {
  text: string;
  intervalMs?: number;
};

const maxFragmentLength = 200;
const firstSequenceNumber = 0;

function bufferFromUtf8(text: string) {
  return new Uint8Array(text.split('').map((c) => c.charCodeAt(0)));
}

export function useAnimatedQREncoder({ text, intervalMs = 200 }: Props) {
  const fragment = useRef<string>('');
  const encoder = useRef<UREncoder | null>(null);
  const messageBuffer = bufferFromUtf8(text);
  // @ts-expect-error - expects Buffer, but Uint8Array works
  const ur = UR.fromBuffer(messageBuffer);
  const instance = new UREncoder(ur, maxFragmentLength, firstSequenceNumber);
  encoder.current = instance;
  // initialize with the first fragment
  // otherwise the QR renders an empty string for the first frame
  fragment.current = encoder.current.nextPart();

  useInterval(
    () => {
      if (encoder.current) {
        fragment.current = encoder.current.nextPart();
      }
    },
    encoder.current ? intervalMs : null,
  );

  return {
    fragment: fragment.current,
    isReady: text.length <= maxFragmentLength || Boolean(encoder),
  };
}
