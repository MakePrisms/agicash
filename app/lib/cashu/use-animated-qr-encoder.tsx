import type { Buffer } from 'node:buffer';
import type { UREncoder } from '@gandlaf21/bc-ur';
import { useEffect, useState } from 'react';
import { useInterval } from '~/lib/use-interval';

type Props = {
  text: string;
  intervalMs?: number;
};

export function useAnimatedQREncoder({ text, intervalMs = 200 }: Props) {
  const [fragment, setFragment] = useState<string>('');
  const [encoder, setEncoder] = useState<UREncoder | null>(null);

  const maxFragmentLength = 150;

  useEffect(() => {
    if (typeof window === 'undefined') return;

    // need to import here because the UREncoder requires a cbor library that only works in the browser
    import('@gandlaf21/bc-ur').then((bcUr) => {
      const { UR, UREncoder } = bcUr;

      const messageBuffer = new Uint8Array(
        text.split('').map((c) => c.charCodeAt(0)),
      );
      const ur = UR.fromBuffer(messageBuffer as Buffer);

      const firstSequenceNumber = 0;

      const encoder = new UREncoder(ur, maxFragmentLength, firstSequenceNumber);
      setEncoder(encoder);

      // Initialize with the first fragment
      setFragment(encoder.nextPart());
    });
  }, [text]);

  useInterval(
    () => {
      if (encoder) {
        setFragment(encoder.nextPart());
      }
    },
    encoder ? intervalMs : null,
  );

  return {
    fragment,
    isReady: text.length <= 150 || Boolean(encoder),
  };
}
