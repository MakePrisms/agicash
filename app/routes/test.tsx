import { Link } from '@remix-run/react';
import { Text } from 'tamagui';
import { TealHeading } from '~/components';

export default function Test() {
  return (
    <div>
      <div>Hey you</div>
      <br />
      <Link to="/">Home</Link>
      <br />
      <Text tag="h1" fontSize={23} color="darkcyan">
        Testing heading
      </Text>
      <TealHeading>it was only a fantasy</TealHeading>
    </div>
  );
}
