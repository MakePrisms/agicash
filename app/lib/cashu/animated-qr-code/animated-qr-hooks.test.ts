import { describe, expect, it, mock } from 'bun:test';
import { renderHook } from '@testing-library/react-hooks';
import { act } from 'react';
import { useAnimatedQRDecoder } from './use-animated-qr-decoder';
import { useAnimatedQREncoder } from './use-animated-qr-encoder';

describe('useAnimatedQREncoder', () => {
  it('should handle short text without encoder', () => {
    const { result } = renderHook(() =>
      useAnimatedQREncoder({ text: 'short text', intervalMs: 100 }),
    );

    expect(result.current.isReady).toBe(true);
    expect(result.current.fragment).toMatchSnapshot();
  });

  it('should initialize encoder for long text', async () => {
    const longText = 'x'.repeat(201);
    const { result } = renderHook(() =>
      useAnimatedQREncoder({ text: longText, intervalMs: 10 }),
    );

    expect(result.current.isReady).toBe(true);
    expect(result.current.fragment).toMatchSnapshot();
  });
});

describe('useAnimatedQRDecoder', () => {
  it('should initialize with empty fragment', () => {
    const onDecode = mock(() => void 0);
    const { result } = renderHook(() =>
      useAnimatedQRDecoder({ fragment: '', onDecode }),
    );

    expect(result.current.progress).toBe(0);
    expect(result.current.error).toBeNull();
    expect(onDecode).not.toHaveBeenCalled();
  });

  it('should handle decoder initialization', async () => {
    const onDecode = mock(() => void 0);
    const { result } = renderHook(() =>
      useAnimatedQRDecoder({
        fragment:
          'ur:bytes/pkhsdycsjkhsehcsishseycsjlhseocsjphseecsjyhseccscxhsencsjyhsemcsihhsetcskshsescsjyceyadtde',
        onDecode,
      }),
    );

    expect(result.current.progress).toBe(0);
    expect(result.current.error).toBeNull();
  });

  it('should handle invalid fragments', async () => {
    const onDecode = mock(() => void 0);
    const { result } = renderHook(() =>
      useAnimatedQRDecoder({ fragment: 'invalid-fragment', onDecode }),
    );

    expect(result.current.error).toBeTruthy();
    expect(onDecode).not.toHaveBeenCalled();
  });
});

describe('end-to-end encoding and decoding', () => {
  it('should successfully encode and decode text through fragments', async () => {
    const originalText = 'Hello World!';
    const onDecode = mock((text: string) => {
      expect(text).toBe(originalText);
    });

    // Set up encoder
    const { result: encoderResult } = renderHook(() =>
      useAnimatedQREncoder({ text: originalText, intervalMs: 10 }),
    );

    // Set up decoder
    const { result: decoderResult } = renderHook(() =>
      useAnimatedQRDecoder({
        fragment: encoderResult.current.fragment,
        onDecode,
      }),
    );

    expect(decoderResult.current.error).toBeNull();
    expect(onDecode).toHaveBeenCalledTimes(1);
  });

  it('should handle long text encoding and decoding', () => {
    const longText = 'x'.repeat(201);
    const onDecode = mock((text: string) => {
      expect(text).toBe(longText);
    });

    // Set up encoder
    const { result: encoderResult } = renderHook(() =>
      useAnimatedQREncoder({ text: longText, intervalMs: 10 }),
    );

    // Simulate receiving multiple fragments
    const { rerender: decoderRerender } = renderHook(
      ({ fragment }) => useAnimatedQRDecoder({ fragment, onDecode }),
      { initialProps: { fragment: encoderResult.current.fragment } },
    );

    // Get next fragment and pass it to decoder
    act(() => {
      for (let i = 0; i < 10; i++) {
        decoderRerender({ fragment: encoderResult.current.fragment });
      }
    });

    expect(onDecode).toHaveBeenCalledTimes(1);
  });
});
