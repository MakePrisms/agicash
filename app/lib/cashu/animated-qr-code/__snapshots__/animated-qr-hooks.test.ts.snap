// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`useAnimatedQREncoder should handle short text without encoder 1`] = `"ur:bytes/pkhsdycsjkhsehcsishseycsjlhseocsjphseecsjyhseccscxhsencsjyhsemcsihhsetcskshsescsjyceyadtde"`;

exports[`useAnimatedQREncoder should initialize encoder for long text 1`] = `"ur:bytes/1-6/lpadamcfaagecytyvymnhphdrlrosohsdycskshsehcskshseycskshseocskshseecskshseccskshsencskshsemcskshsetcskshsescsksidehdycsksidehehcsksideheycsksideheocsksideheecsksideheccsksidehencsksidehemcsksidehetcsksidehescsksideydycsksideyehcsksideyeycsksideyeocsksideyeecsksideyeccsksideyencsksideyemcsksideyetcsksideyescsksideodycsksideoehcsksideoeycsksideoeocsksideoeecsksideoeccsksideoencsksideoemcsksidbnhkisdy"`;
